gcp_variant_transforms/beam_io/vcfio.py

```diff
@@ -207,6 +209,168 @@ def __repr__(self):
         [str(s) for s in [self.name, self.genotype, self.phaseset, self.info]])
 
 
+
+class _VcfSink(filebasedsink.FileBasedSink):
+  """A sink for writing VCF files."""
+
+  class _ToVcfRecordCoder(coders.Coder):
+    """Coder for encoding :class:`Variant` objects as VCF text lines."""
+
+    def _is_alternate_phaseset(self, call):
+      return call.phaseset and call.phaseset != DEFAULT_PHASESET_VALUE
+
+    def _format_value(self, value):
+      """Formats a Variant field value to be output to a VCF file.
+
+      Replaces None and [] values with MISSING_FIELD_VALUE and is called
+      recursively for all values in lists.
+
+      Args:
+        value: The value that will be formatted for output to VCF.
+      Returns:
+        A string of the `value` formatted for output to VCF.
+      """
+      def _field_missing(value):
+        """Returns True for empty list, string or None and False otherwise."""
+        return not value and value != 0
+
+      if isinstance(value, list):
+        return ','.join([self._format_value(x) for x in value])
+      return MISSING_FIELD_VALUE if _field_missing(value) else str(value)
+
+    def _get_variant_info(self, variant):
+      info_list = []
+      # Set END in info if it doesn't match start+len(reference_bases). This is
+      # usually the case for non-variant regions.
+      if (variant.start is not None
+          and variant.reference_bases
+          and variant.end
+          and variant.start + len(variant.reference_bases) != variant.end):
+        info_list.append('END=%d' % variant.end)
+      # Set all other fields of info.
+      for k, v in variant.info.iteritems():
+        if v.data is True:
+          info_list.append(k)
+        else:
+          info_list.append('%s=%s' % (k, self._format_value(v.data)))
+      return info_list
+
+    def _get_variant_format_keys(self, variant):
+      if not variant.calls:
+        return []
+
+      format_keys = [GENOTYPE_FORMAT_KEY]
+      for call in variant.calls:
+        # If any calls have a set phaseset that is not `DEFAULT_PHASESET_VALUE`,
+        # the key will be added to the format field.
+        if self._is_alternate_phaseset(call):
+          format_keys.append(PHASESET_FORMAT_KEY)
+        format_keys.extend([k for k in call.info])
+
+      # Sort all keys and remove duplicates after GT
+      format_keys[1:] = sorted(list(set(format_keys[1:])))
+
+      return format_keys
+
+    def _get_variant_calls(self, variant, format_keys):
+      calls = []
+      for call in variant.calls:
+        call_info = []
+        # GT will always be the first format key.
+        genotype_field = MISSING_FIELD_VALUE
+        if MISSING_GENOTYPE_VALUE not in call.genotype:
```

Comment:
@arostamianfar: Should these be applied element wise on the list? Like could genotype be [1, -1] in which case we would want to output "1/." instead?